### DISCUSSION TITLE
how to get the unique products from library enumeration?

### DISCUSSION THREAD

**Discussion by rczerminski-valo** [rdkit/rdkit#4805] [Ideas]
According to `https://rdkit.org/docs/GettingStartedInPython.html`
"You can use canonical smiles and a python dictionary to get the unique products" (from library enumeration)

**Is there a better way to do this?**

For example, if I enumerate library and then want to calculate some properties for the products I would rather do it directly using `<rdkit.Chem.rdchem.Mol>` objects instead of converting mols to smiles to get unique set and then back to rdkit mol. objects for properties calculation.

**Comment by greglandrum** [comment]
Hi @rczerminski-valo
I wouldn't convert the molecule to SMILES and then convert them back.
The example in the documentation uses a dictionary based on the SMILES just to keep track of what's already been seen, but it keeps the original product molecules.

Here's an adaptation of that example:
```
In [17]: rxn = AllChem.ReactionFromSmarts('[C:1]=[C:2].[C:3]=[*:4][*:5]=[C:6]>>[C:1]1[C:2][C:3][*:4]=[*:5][C:6]1')

In [18]: ps = rxn.RunReactants((Chem.MolFromSmiles('OC=C'), Chem.MolFromSmiles('C=CC(N)=C')))

In [19]: uniqps = {}

In [20]: for p in ps:
    ...:     Chem.SanitizeMol(p[0])
    ...:     uniqps[Chem.MolToSmiles(p[0])] = p[0]
    ...: 

In [21]: uniqps.keys()
Out[21]: dict_keys(['NC1=CCCC(O)C1', 'NC1=CCC(O)CC1'])

In [22]: uniqps.values()
Out[22]: dict_values([<rdkit.Chem.rdchem.Mol object at 0x7ff10776e7b0>, <rdkit.Chem.rdchem.Mol object at 0x7ff120ba2190>])
```

**Comment by greglandrum** [comment]
If you just want to calculate properties and not store the molecules you could do something like this:
```
In [23]: from rdkit.Chem import Descriptors

In [24]: props = {}

In [25]: ps = rxn.RunReactants((Chem.MolFromSmiles('OC=C'), Chem.MolFromSmiles('C=CC(N)=C')))

In [26]: for p in ps:
    ...:     Chem.SanitizeMol(p[0])
    ...:     smi = Chem.MolToSmiles(p[0])
    ...:     if smi in props:
    ...:         continue
    ...:     props[smi] = [Descriptors.MolLogP(p[0]),Descriptors.MolWt(p[0]),Descriptors.NumRotatableBonds(p[0])]
    ...: 
```

**Comment by cottonkiet** [comment]
A similar question related to this. Dissociation reactions produce more than one products (for example, the back reaction of the above example). What would be a good way to get unique pair of products and store their smiles properly? Thanks a lot!
```
rxn = AllChem.ReactionFromSmarts('[C:1]1[C:2][C:3][*:4]=[*:5][C:6]1>>[C:1]=[C:2].[C:3]=[*:4][*:5]=[C:6]')
ps = rxn.RunReactants([Chem.MolFromSmiles('NC1=CCCC(O)C1')])
print(len(ps)) # result is 2
Chem.MolToSmiles(ps[0][0]) 
Chem.MolToSmiles(ps[0][1])
```

**Reply by greglandrum** [reply]
The best way to "uniquify' the products of a reaction is to sanitize the molecules and then generate canonical SMILES.
